src/validops.cu

```diff
@@ -85,52 +112,54 @@ gdf_error gdf_count_nonzero_mask(gdf_valid_type const * masks, int num_rows, int
   const size_t num_masks = gdf_get_num_chars_bitmask(num_rows);
 
   // Number of 4 byte types in the validity bit mask 
-  const size_t num_masks32 = num_masks / RATIO;
+  size_t num_masks32 = num_masks / RATIO;
 
   // If the total number of masks is not a multiple of the RATIO 
   // between the original mask type and the 4 byte masks type, then 
   // these "remainder" masks cannot be proccessed in the transform_reduce
   // and must be handled separately
   cudaStream_t copy_stream;
-  const size_t num_remainder_masks = num_masks % RATIO;
+  size_t num_remainder_masks = num_masks % RATIO;
+
+  // If there are no remainder masks, the last mask still
+  // needs to be handled separately
+  if(0 == num_remainder_masks){
+    num_remainder_masks = RATIO;
+    num_masks32--;
+  }
+
   std::vector<gdf_valid_type> remainder_masks(num_remainder_masks); 
   if(remainder_masks.size() > 0)
   {
-
     CUDA_TRY(cudaStreamCreate(&copy_stream));
 
     // Copy the remainder masks to the host
     // FIXME: Is this endian safe?
-    const gdf_valid_type * first_remainder_mask = (masks + num_masks) - num_remainder_masks;
+    const gdf_valid_type * first_remainder_mask = &(masks[num_masks - num_remainder_masks]);
+
     CUDA_TRY( cudaMemcpyAsync(remainder_masks.data(), 
```

Comment:
Next optimization might be to leave the remainder masks on the device and count them there...  Even do it all in the same kernel now that you aren't using Thrust!